Test for a no-media call.

alvestrand opened the pull request:
This test verifies that a connection is established for a call where
no media is added, but the initiator offers to receive video.

The iceconnectionstate = connected indicator is used to determine
that the connection is established.

(This test submission is also a test on the review process for submitting new tests. Please give feedback as appropriate.)

alvestrand committed:
Test for a no-media call.

This test verifies that a connection is established for a call where
no media is added, but the initiator offers to receive video.

The iceconnectionstate = connected indicator is used to determine
that the connection is established.

alvestrand committed:
Comment update: Noted bug number for spec issue

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+  <script src="/resources/testharnessreport.js"></script>
+  <script src="/common/vendor-prefix.js"
+          data-prefixed-objects=
+              '[{"ancestors":["navigator"], "name":"getUserMedia"},
```
getUserMedia is not used, so no need to enable its prefixing

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+                {"ancestors":["window"], "name":"RTCSessionDescription"},
+                {"ancestors":["window"], "name":"RTCIceCandidate"}]'
+          data-prefixed-prototypes=
+               '[{"ancestors":["HTMLMediaElement"],"name":"srcObject"}]'>
```
srcObject is not used, so no need to enable its prefixing

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+  </script>
+  <script type="text/javascript">
+  var test = async_test('Can set up a basic WebRTC call with no data.',
+                        {timeout: 5000});
```
why the 5000ms timeout? it's worth documenting the reason with a comment I think

alvestrand replied:
Is there any guidance on this? We know that most bugs will cause a timeout, but timeouts just above the expected runtime causes flakiness, so a timeout "well beyond reasonable" seems like the right thing to put here.

dontcallmedom replied:
see http://testthewebforward.org/docs/testharness-library.html#harness-timeout

alvestrand replied:
Set to default.

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+
+  // This function starts the test.
+  test.step(function() {
+    gFirstConnection = new RTCPeerConnection(null, null);
```
why 2 params to RTCPeerConnection? (similar comment on line 119)

alvestrand replied:
Works with one, so changing.

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+                                                  sdp: offerSdp });
+    gSecondConnection.setRemoteDescription(parsedOffer);
+
+    gSecondConnection.createAnswer(onAnswerCreated,
```
the spec promotes the usage of the promise-based version of the interface; the test should use it too.

alvestrand replied:
TODO added.

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+
+  // Note: the ice candidate handlers are special. We can not wrap them in test
+  // steps since that seems to cause some kind of starvation that prevents the
+  // call of being set up. Unfortunately we cannot report errors in here.
```
is this because a bug in the spec? in the implementation? somewhere else? ideally, this should link to a relevant bug report to know if and when that piece can get wrapped in a proper test step.

alvestrand replied:
This is copied from the test I based it on, I don't know what it means. What does "wrap in a step" mean, and what is different when you do? Pointer?

dontcallmedom replied:
"wrap in a test step" means using the same test.step_func() wrapper as for the other event handlers. Doing so means that errors that are encountered in that step will be reported; it also lets you use assert_* calls in the body of the handler. See http://testthewebforward.org/docs/testharness-library.html#asynchronous-tests

alvestrand replied:
They worked when wrapping them, so the error has gone away.

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,132 @@
+
+    // The offerToReceiveVideo is necessary and sufficient to make
+    // an actual connection.
+    gFirstConnection.createOffer(onOfferCreated, failed('createOffer'),
```
should use promises

alvestrand replied:
do we have a promises based shim already? we sure don't have native support yet.

dontcallmedom replied:
the tests are supposed to set what implementations aspire to match (i.e. the spec), rather than reflect their current limitations :)

dontcallmedom commented:
I've taken a first pass at commenting on the test case; also, the test times out in firefox (whereas it passes in Chrome) — not sure why yet.

alvestrand committed:
Review comments from Dom.

Used wrappers for the rest of callbacks, added TODO about using
a promise-based API, addressed other comments.

alvestrand commented:
PTAL

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,131 @@
+    // Note: This test should have been as below, but second connection
+    // does not reach "completed". This is noted as an issue in the spec.
+    // https://github.com/w3c/webrtc-pc/issues/224
+    if (gFirstConnection.iceConnectionState == 'completed' &&
```
since https://github.com/w3c/webrtc-pc/issues/224 is now closed (or about to be), that needs an update; if "completed" && "completed" is the expected outcome, we should only keep that version.

(I've since determined that the timeout in firefox I reported separately is due to it being stuck in "connected" / "connected" apparently)

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,131 @@
+  <script src="/resources/testharnessreport.js"></script>
+  <script src="/common/vendor-prefix.js"
+          data-prefixed-objects=
+              '[{"ancestors":["window"], "name":"RTCPeerConnection"}]'
```
having the prefixing on RTCSessionDescription and RTCIceCandidate was useful for running the test in Firefox; could you reinstate those?

alvestrand committed:
More updates.

Test now runs in Firefox, except that Firefox doesn't implement the
oniceconnectionstatechange callback.

There is a display on the test page for the state info, and the comment
explaining the reason for early test termination has been expanded.

alvestrand commented:
PTAL.

Restored the prefixing, and also added a little more polish (as well as removed the video display tags that we don't use). One more Firefox issue found and bypassed (it doesn't queue createAnswer behind setLocalDescription).

Also displayed what the status is at test end.

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,146 @@
+  <script src="/common/vendor-prefix.js"
+          data-prefixed-objects=
+              '[{"ancestors":["window"], "name":"RTCPeerConnection"},
+                {"ancestors":["window"], "name":"RTCSessionDescription"}]'
```
still misses one for RTCIceCandidate for Firefox sake

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,146 @@
+
+  var onOfferCreated = test.step_func(function(offer) {
+    // TODO: Switch to promise-based interface.
+    gFirstConnection.setLocalDescription(offer, null,
```
null is not an acceptable argument for setLocalDescription (and thus fails in Firefox); you need function(){} there
(likewise in line 57 and 67)

dontcallmedom commented on webrtc/no-media-call.html:
```diff
@@ -0,0 +1,146 @@
+                        + '<br>Second: ' + gSecondConnection.iceConnectionState;
+    // Note: All these combinations are legal states indicating that the
+    // call has connected. All browsers should end up in completed/completed,
+    // but as of this moment, we've chosen to terminate the test early.
```
is this a TODO? or is that reasonable thing to test for anyway?

alvestrand replied:
It means the test is currently valid, in that it checks a place we expect the state machine to arrive at, but there is a state it does not test for.

I'll add a TODO.

dontcallmedom commented:
we're nearly there :)

alvestrand committed:
Ensure test passes in Firefox.

alvestrand commented:
Great - I now have a test that works in Firefox 37.0.2 too - the last fix did it!
Wonderful!
